### PR TITLE
Remove wildcard imports from the codebase

### LINKED ISSUE #11852: Remove usage of Kotlin wildcard imports on the project Codestyle file
I agree. Although this one was added automatically by the IDE.

Default for Kotlin is 3 imports to substitute to `*` (https://stackoverflow.com/questions/49870306/disable-wild-cart-import-in-intellij-android-studio-in-kotlin)

It can be a good first issue to change it on the project codestyle file (https://github.com/ankidroid/Anki-Android/blob/main/.idea/codeStyles/Project.xml)

_Originally posted by @BrayanDSO in https://github.com/ankidroid/Anki-Android/pull/11803#discussion_r921690661_

### PR DESCRIPTION
**Pull request by andriusezerskis**
## Purpose / Description
Removed all the wildcard imports from the codebase and changed linter rules to not allow them anymore

## Fixes
* Fixes #11852 

## Approach
No more wildcard imports

## How Has This Been Tested?
I just ran the linter and unit tests before pushing to remote

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by david-allison** [comment]
CI is failing

**Comment by david-allison** [comment]
Hi, in future: 

* Could you rebase + force push, rather than making a new PR - we want a history of the code review
* Rebase onto `main` rather than using merge commits in the history

**Comment by andriusezerskis** [comment]
I'm sorry for the 2 PRs, I won't do that again. So I combed through the failed stuff for the PR and there's basically those 3 errors a bunch of times. I don't really understand how the first one affects my PR, since the other PRs seem to be fine with the checks. The second one seems to be a result of the deprecated preference thing again and the third one, I checked with the imports to see if I had messed up something, but it looks like it does do an assert with the `FileNotFoundException`.
> Task :AnkiDroid:lintPlayDebug FAILED
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/res/values/10-preferences.xml:221: Error: The resource R.string.deck_conf_cram_filter appears to be unused [UnusedResources]
    <string name="deck_conf_cram_filter" maxLength="41">Filter</string>
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~

com.ichi2.anki.FilteredDeckOptionsTest > canOpenDeckOptions[emulator-5554 - 11] FAILED 
	java.lang.NullPointerException: null cannot be cast to non-null type androidx.preference.ListPreference
	at com.ichi2.anki.FilteredDeckOptions.buildLists(FilteredDeckOptions.kt:323)
	

DirectoryTest > [CompatV26] > has_files_throws_if_file_no_longer_exists[CompatV26] FAILED
    org.opentest4j.AssertionFailedError: Expected an exception of class java.io.FileNotFoundException to be thrown, but was java.nio.file.NoSuchFileException: /tmp/10787257164438226601
        at app//org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:42)
        at app//org.junit.jupiter.api.Assertions.fail(Assertions.java:150)
        at app//kotlin.test.junit5.JUnit5Asserter.fail(JUnitSupport.kt:63)
        at app//kotlin.test.AssertionsKt__AssertionsImplKt.checkResultIsFailure(AssertionsImpl.kt:30)
        at app//kotlin.test.AssertionsKt.checkResultIsFailure(Unknown Source)
        at app//com.ichi2.anki.model.DirectoryTest.has_files_throws_if_file_no_longer_exists(DirectoryTest.kt:98)

        Caused by:
        java.nio.file.NoSuchFileException: /tmp/10787257164438226601
            at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
            at java.base/sun.nio.fs.UnixException.asIOException(UnixException.java:115)
            at java.base/sun.nio.fs.UnixFileSystemProvider.newDirectoryStream(UnixFileSystemProvider.java:477)
            at java.base/java.nio.file.Files.newDirectoryStream(Files.java:481)
            at com.ichi2.compat.CompatV26.newDirectoryStream(CompatV26.kt:93)
            at com.ichi2.compat.CompatV26.contentOfDirectory(CompatV26.kt:103)
            at com.ichi2.compat.Compat.hasFiles(Compat.kt:159)
            at com.ichi2.anki.model.Directory.hasFiles(Directory.kt:50)
            ... 1 more

**Comment by david-allison** [comment]
Remove the merge commits from the history and it will be much easier to take a look

**Comment by david-allison** [comment]
Number 3:

```patch
Index: AnkiDroid/src/main/java/com/ichi2/compat/CompatV26.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/compat/CompatV26.kt b/AnkiDroid/src/main/java/com/ichi2/compat/CompatV26.kt
--- a/AnkiDroid/src/main/java/com/ichi2/compat/CompatV26.kt	(revision daf3ad6bf727c9e5a3ab650dcc4dd1b15c9b090a)
+++ b/AnkiDroid/src/main/java/com/ichi2/compat/CompatV26.kt	(date 1721056646490)
@@ -31,6 +31,7 @@
 import java.nio.file.DirectoryIteratorException
 import java.nio.file.DirectoryStream
 import java.nio.file.Files
+import java.nio.file.NoSuchFileException
 import java.nio.file.Path
 import java.nio.file.Paths
 import java.nio.file.StandardCopyOption
```

**Comment by david-allison** [comment]
Did you expand the imports using the IDE, or manually?

**Comment by david-allison** [comment]
With apologies, you're going to want to look into how to do this in an automated/semi-automated fashion to ensure that we're not introducing bugs via changing imports

* Manual changes have introduced bugs
* This is a huge change, and I wouldn't expect anyone to test all affected code paths
* A number of these bugs were picked up by our test suite
* I am not confident in the coverage of our test suite
* So there's a fair likelihood that uncaught bugs will have been introduced
  * Especially around `java.io` vs `java.nio`

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt [hunk 1]
```diff
@@ -86,26 +134,45 @@ import com.ichi2.annotations.NeedsTest
...
+import com.ichi2.utils.title
 import com.squareup.seismic.ShakeDetector
 import kotlinx.coroutines.Job
 import kotlinx.coroutines.runBlocking
 import timber.log.Timber
-import java.io.*
+import java.io.File
+import java.io.UnsupportedEncodingException
 import java.net.URLDecoder
-import java.util.*
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt:108]: Confirmed that this was not needed

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 2]
```diff
@@ -952,7 +990,7 @@ open class CardBrowser :
      * @return `false` if the user may proceed; `true` if a warning is shown due to being in [NOTES]
      */
     private fun warnUserIfInNotesOnlyMode(): Boolean {
-        if (viewModel.cardsOrNotes != NOTES) return false
+        if (viewModel.cardsOrNotes != CardsOrNotes.NOTES) return false
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:993]: Open up the Android Studio Menu when the cursor is on the dot (Alt + Enter on my mac) and import the CardsOrNotes.NOTES member, so this is not a code change

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 3]
```diff
@@ -149,6 +207,7 @@ import java.io.File
 import java.lang.Runnable
 import kotlin.time.Duration.Companion.milliseconds
 import kotlin.time.measureTimedValue
+import com.ichi2.anki.BuildConfig as AnkiBuildConfig
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:210]: Shouldn't be necessary - try and keep each refactoring commit to only do one thing.

I would accept this in a separate PR/commit due to BuildConfig ambiguity

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 4]
```diff
@@ -1577,9 +1636,9 @@ open class DeckPicker :
             // There the "lastVersion" is set, so that this code is not reached again
             if (VersionUtils.isReleaseVersion) {
                 Timber.i("Displaying new features")
-                val infoIntent = Intent(this, Info::class.java)
-                infoIntent.putExtra(Info.TYPE_EXTRA, Info.TYPE_NEW_VERSION)
-                showNewVersionInfoLauncher.launch(infoIntent)
+                val infoIntentStart = Intent(this, Info::class.java)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:1639]: Each refactoring commit should do one thing, revert this please

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsArrayAdapter.kt [hunk 5]
```diff
@@ -35,7 +40,9 @@ import java.util.TreeSet
 /**
  * @param tags A reference to the [TagsList]
  */
-class TagsArrayAdapter(private val tags: TagsList, private val resources: Resources) : RecyclerView.Adapter<TagsArrayAdapter.ViewHolder>(), Filterable {
+class TagsArrayAdapter(private val tags: TagsList, private val resources: Resources) :
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsArrayAdapter.kt:43]: revert the spacing change

It's correct, just not in scope for a refactoring across the codebase

File: AnkiDroid/src/test/java/com/ichi2/anki/servicelayer/scopedstorage/MoveFileTest.kt [hunk 6]
```diff
@@ -224,7 +233,7 @@ class MoveFileTest(private val attemptRename: Boolean) : RobolectricTest(), Oper
         val source = addUntrackedMediaFile("hello-oo", listOf("hello.txt"))
         val destinationFile = source.file
 
-        val ex = assertFailsWith<EquivalentFileException> {
+        val ex = assertFailsWith<MigrateUserData.EquivalentFileException> {
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/anki/servicelayer/scopedstorage/MoveFileTest.kt:236]: revert

**Review by david-allison — CHANGES_REQUESTED** [review]
Monster effort. Only nitpicks, thanks so much!!!!!

---

I would like to rebase merge this PR, rather than squash merging.

If we're rebasing, each commit should (ideally) do one logical thing, compile, and tell a story. Any fixup commits should definitely be squashed - people looking at the history would see these as noise.

It's important that the editor config change is in a separate commit, to separate it out from the 'mass' change

Please rebase + squash + force-push the follow-up commits which contain the import-related fixes, so this contains a clean history and can be merged

This should leave 2 commits:

* update editorConfig
* update the imports

Use your better judgment if you can see a third relevant commit

----

Once the commit hashes are in `main`, a follow-up Pull Request would be to add the second commit's ref to: https://github.com/ankidroid/Anki-Android/blob/main/.git-blame-ignore-revs

**Review by david-allison — APPROVED** [review]
Excellent, thanks so much!!!

**Comment by andriusezerskis** [comment]
Thanks for your patience and advice !! I've learned a lot and will definitely try to contribute more 😄

**Review by lukstbit — APPROVED** [review]
LGTM, thanks!

@david-allison Do we merge? This has the potential to create conflicts in a lot of PRs like the Gsoc ones.

**Comment by BrayanDSO** [comment]
I'm going to merge it. It's easier to rebase the other PRs than this one, and we won't ever get to a point where this doesn't create a lot of conflicts.

**Comment by david-allison** [comment]
Agreed - thanks for merging!

And thank you so much @andriusezerskis 

I'll open a follow-up issue for the commit hash, but feel free to throw it back to us. It's a really quick (but useful) maintenance job

**Pull request merged**